doc/spec_TransferElement.dox

```diff
@@ -213,9 +232,50 @@ This documnent is currently still **INCOMPLETE**!
+  - 12.4 A TransferGroup is becomes read-only (cf. TransferGroup::isReadOnly()), if at least one of the member TransferElements is read-only.
+  - 12.5 Adding the same TransferElement twice is not allowed (logic_error).
+  - 12.6 Adding one TransferElement to two different TransferGroups is not allowed (logic_error).
+  - 12.7 Adding TransferElements with AccessMode::wait_for_new_data is not allowed (logic_error).
+  - 12.8 Adding a copy of the a TransferElement which is already in the group (returned by two different calls to Device::getZzzRegisterAccessor() resp. DeviceBackend::getRegisterAccessor() with the same arguments) will replace one of the identical TransferElement with a decorator which copies the data from the other.
+    - 12.8.1 This avoids an unnecessary transfer of identical data.
+    - 12.8.2 This is only possible, if the TransferElement is added by passing a TransferElementAbstractor to TransferGroup::addAccessor().
+  - 12.9 Operations executed via TransferGroup::read() and TransferGroup::write():
+    - 12.9.1 first call preXxx() on all high-level elements (cf. \ref transferElement_B_12_1_2_1 "12.1.2.1"),
+    - 12.9.2 next call the corresponding xxxTransferYyy() on all low-level elements (cf. \ref transferElement_B_12_1_2_2 "12.1.2.2"), and
+    - 12.9.3 finally call postXxx() on all high-level elements.
+  - 12.10 The TransferGroup implements a similar exception handling in TransferGroup::read() and TransferGroup::write() as their TransferElement counterparts. This especially means:
+    - 12.10.1 Before preXxx() is called, the TransferGroup checks all pre-conditions (device opened and operation allowed) to make sure no ChimeraTK::logic_error can occur (other exceptions are anyway not allowed in preXxx).
+      - \anchor transferElement_B_12_10_1_1 12.10.1.1 If any ChimeraTK::runtime_error occurs while checking the pre-conditions, it is immediately passed on to the application.
+      - \anchor transferElement_B_12_10_1_2 12.10.1.2 Any pre-condition which is not met will immediately result in a thrown ChimeraTK::logic_error
+      - 12.10.1.3 The exceptions from \ref transferElement_B_12_10_1_1 "12.10.1.1" and \ref transferElement_B_12_10_1_2 "12.10.1.2" are not prioritised by type. The first detected error throws.
+    - 12.10.2 ChimeraTK::runtime_error exceptions thrown by the low-level elements are propagated to the corresponding high-level elements so they are seen by their postXxx() functions (c.f. \ref transferElement_B_6_3 "6.3")
+    - 12.10.3 postXxx() is called for all high level elements, even if some postXxx() throw exceptions (c.f. \ref transferElement_B_5 "5").
+      - 12.10.3.1 If exceptions have been caught from any postXxx(), an exception is thrown after all of them have been executed.
```

Comment:
Proposal for 12.10.3.1
```
* 12.10.3.1 All exception messages of the postXxx() calls are collected.
```

Reply:
Better:
- 12.10.3.1.1 The thrown exception will contain all messages from the caught exceptions.

Reply:
Actually, I think this is a bad idea. We should merely keep the first message. Otherwise we will get in the vast majority of cases very very long error messages which are repeating the same error many times. Probably it will be completely unreadable and also create displaying issues due to too long strings. The chance that there is a real value in the additional information which can actually be found is rather slim.

TransferGroups may have hundreds of variables of usually very few devices. Most of the time all (single device) or a good fraction (few devices) of the variables will have the same error simultaneously. Even if two devices fail at the same time, who is supposed to find the actually different message in that pile of messages?

We could invent some sophisticated mechanism to filter duplicates, but this may be spoiled by the actually good practise of including the variable name in the message (-> we should probably even require this from all backends, helps debugging a lot). It's not worth doing this. Just keep the first message and that's it.

Reply:
Hence:

> * 12.10.3.1.1 The thrown exception will contain the message of the first caught exception.

Reply:
You are right. But it is not the first caught exception. I'll explain why:

* What actually will happen with the new requirement of setException is that after the first transfer gave a useful message, all others will say "follow-up error", which we don't want.
* So we have to get the message from the first transfer which has seen an exception. But it is not thrown, it happens under the hood. We only see that _activeException is not nullptr.
* We cannot get the first thrown exception from the post reads. They might happen in a different order than the transfers.

Reply:
Yes it is the first caught/thrown exception. The original throwing happens in the transfer. Everything else is just rethrowing. How the TransferGroup organises itself to make sure that actually the originally first error is rethrown is an implementation detail.

So maybe it is clearer like this:

> 12.10.3.1.1 The thrown exception will contain the message of the first thrown exception.

Reply:
I don't know if what I am writing is helpful, because it it not only about this point. Here is my summary of all the discussions, without distinction to which point it belongs:

* boost::bad_numeric_cast is not treated
* boost::thread_interrupted cannot occur
* preConditons check that no ChimeraTK::logic_error can occur in preXxx, as described
* => only runtime_errors are left, can only primarily happen on transfer phase
* exceptions in the transfer phase are delayed,  so all postXxx are called with the correct exception
* If an exception is seen in the transfer phase, all postRead are called with UpdateDataBuffer = false
* All re-thrown exceptions from the postXxx-phase are discarded.
* The first detected exception from the transfer phase is re-thrown after all postXxx have been executed.

Reply:
numeric_cast should be treated, so that all postReads are called. We just don't have to care about priorities because stuff will natively come in a well-defined order. Just throwing the first error (after all postXxx have been executed) is enough.